### DISCUSSION TITLE
How to move entities between layouts

### DISCUSSION THREAD

**Discussion by sunblog3** [mozman/ezdxf#1076] [Q&A] [answered]
```python
Found non-unique entity handle #559AC, data validation is required.
Found non-unique entity handle #559AD, data validation is required.
Found non-unique entity handle #559AE, data validation is required.
Found non-unique entity handle #559AF, data validation is required.
Found non-unique entity handle #559B0, data validation is required.
Found non-unique entity handle #559B1, data validation is required.
Found non-unique entity handle #559B2, data validation is required.
```
How should I clear the alert, or write a program to skip this

**Comment by mozman** [comment]
You can't do very much (with ezdxf), this is a broken file because handles have to be unique. If its a DXF R12 file you can remove the handles by:

    ezdxf strip --handles <your.dxf>

If its a newer DXF version you can try a real CAD application to recover the file.

**Reply by sunblog3** [reply]
No error is reported before the block is created, but an error is reported after the block is created
```python
import ezdxf

filename = os.path.basename("test.dxf")
stem, suffix = os.path.splitext(filename)
doc = ezdxf.readfile("test.dxf")
msp = doc.modelspace()
entities = [entity for entity in msp]

# Check whether the file is damaged
print(ezdxf.readfile("test.dxf"))

# Create a new block
block = doc.blocks.new(name=stem)
for entity in entities:
    #  entity to the block
    block.add_entity(entity)
msp.add_blockref(name=stem, insert=(0, 0))
# Save the DXF file
doc.saveas("test_block.dxf")
# Load file

print(ezdxf.readfile("test_block.dxf"))
```
Error message:
```python
<ezdxf.document.Drawing object at 0x000001FCA8818C50>
Found non-unique entity handle #2B8, data validation is required.
Found non-unique entity handle #2B9, data validation is required.
....
<ezdxf.document.Drawing object at 0x000001FCA8B93810>
```

[demo.zip](https://github.com/mozman/ezdxf/files/15281605/demo.zip)

**Reply by mozman** [reply]
You are loading DXF files from unreliables sources at least use the [recover](https://ezdxf.mozman.at/docs/drawing/recover.html) module to load the drawing.

**Reply by mozman** [reply, marked as answer]
It's hard to stop users from doing stupid things when the data format allows for a lot of stupid things.

In this case you are trying to assign entities from modelspace to a new block, but each entity can only be assigned to one layout. The `add_entity()` method is a low level tool and should only be used if you know what you are doing.

There exist a method for moving entities to between layouts: [BaseLayout.move_to_layout()](https://ezdxf.mozman.at/docs/layouts/layouts.html#ezdxf.layouts.BaseLayout.move_to_layout)

And no, there is no way to prevent this from happening in the first place, as adding an entity to a second layout can be an intermediate step in a larger workflow:
- `ezdxf` is a tool for programmers, not an end-user application!
- you need to understand the basics of the data format
- read the documentation e.g. of `add_entity()`: 
  - Add an existing DXFGraphic entity to a layout, but be sure to unlink ([unlink_entity()](https://ezdxf.mozman.at/docs/layouts/layouts.html#ezdxf.layouts.BaseLayout.unlink_entity)) entity from the previous owner layout. Adding entities from a different DXF drawing is not supported.
